### PR TITLE
<Box /> - Component

### PR DESCRIPTION
**Pull request by barakplasma**
provides a more generic Box element which can replace vBox and hBox, with a prop for vertical or horizontal alignment.

### CODE REVIEW

File: packages/wix-ui-core/stories/index.tsx [hunk 1]
```diff
@@ -52,6 +53,33 @@ storiesOf('Components', module)
         <div>c</div>
     </HBox>
   ))
+  .add('Box', () => (
```
Comment (lbelinsk) [inline packages/wix-ui-core/stories/index.tsx]: You don't need to docoment VBox & HBox anymore

File: packages/wix-ui-core/src/components/Box/Box.driver.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  let domTestkit = null;
+  if (componentInstance) {
+    domTestkit =  new DomTestkit({componentId: componentInstance.id} );
+  }
+
+  return {
+    /** check if element exists */
+    exists: () => !!element,
+    /** return box flex direction value */
+    getFlexDirection: () => domTestkit.getCssValue({className: 'boxRoot', property: 'flex-direction'}),
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/Box.driver.ts]: Sorry, but today we learned that there is a better practice to test this, see here:
https://github.com/wix/wix-ui/pull/117/files#diff-ea0b6f1a99f1c8d26a4751f6620f1e60R6

Please update it and see that the tests still passes

File: packages/wix-ui-core/src/components/Box/Box.driver.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    /** return box item alignment value */
+    getAlignment: () => domTestkit.getCssValue({className: 'boxRoot', property: 'align-items'})
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/Box.driver.ts]: Please add tests for `spacing ` prop as well

File: packages/wix-ui-core/src/components/Box/Box.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  it('should render the passed children', () => {
+    wrapper = mount(<Box theme={{boxType: 'vertical'}}><div>1</div></Box>, {attachTo: document.createElement('div')});
+    expect(wrapper.html()).toContain('<div>1</div>');
+  });
+});
+
+describe('Box styling', () => {
+  const createDriver = createDriverFactory(boxDriverFactory);
+
+  it('vBox should use flex direction column', () => {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/Box.spec.tsx]: no vbox, only box

File: packages/wix-ui-core/src/components/Box/Box.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    const driver = createDriver(<Box theme={{boxType: 'vertical'}}><div>1</div></Box>);
+    expect(driver.getFlexDirection()).toBe('column');
+  });
+  it('hBox should use flex direction row', () => {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/Box.spec.tsx]: no hbox

File: packages/wix-ui-core/src/components/Box/Box.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    const driver = createDriver(<Box theme={{boxType: 'horizontal'}}><div>1</div></Box>);
+    expect(driver.getFlexDirection()).toBe('row');
+  });
+  it('hBox should have alignment bottom by default', () => {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/Box.spec.tsx]: no hbox

File: packages/wix-ui-core/src/components/Box/theme.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+export type BoxTheme = VBoxTheme | HBoxTheme;
+
+export type VBoxTheme = {
+  boxType: 'vertical'
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/theme.ts]: boxType is not mandatory.

Moreover, maybe a nicer API will be some boolean prop? something like `vertical`: bool which is false by default

File: packages/wix-ui-core/src/components/Box/Box.spec.tsx (hunk 4 above)
Comment (barakplasma) [reply packages/wix-ui-core/src/components/Box/Box.spec.tsx]: I don't understand, we are trying to test <Box type=vertical /> here, not vBox the old component

File: packages/wix-ui-core/src/components/Box/Box.spec.tsx (hunk 5 above)
Comment (barakplasma) [reply packages/wix-ui-core/src/components/Box/Box.spec.tsx]: I don't understand, we are trying to test <Box type=horizontal /> here, not hBox the old component

File: packages/wix-ui-core/src/components/Box/theme.ts (hunk 7 above)
Comment (barakplasma) [reply packages/wix-ui-core/src/components/Box/theme.ts]: So horizontal boxes should be the default, and a flag with make it vertical instead? Will do.

File: packages/wix-ui-core/stories/index.tsx (hunk 1 above)
Comment (barakplasma) [reply packages/wix-ui-core/stories/index.tsx]: removed VBox and HBox documentation - 2785701ab8dfb2be43692d952e80eae216f46579

File: packages/wix-ui-core/src/components/Box/theme.ts (hunk 7 above)
Comment (barakplasma) [reply packages/wix-ui-core/src/components/Box/theme.ts]: done e2c3074

File: packages/wix-ui-core/src/components/Box/Box.driver.ts (hunk 2 above)
Comment (barakplasma) [reply packages/wix-ui-core/src/components/Box/Box.driver.ts]: updated with `window.getComputedStyle()`

File: packages/wix-ui-core/stories/index.tsx (hunk 1 above)
Comment (barakplasma) [reply packages/wix-ui-core/stories/index.tsx]: typescript warning was breaking the build: fixed in https://github.com/wix/wix-ui/pull/126/commits/31e929bea4a6584e969c223bc8433a78e9285e64

**Comment by barakplasma** [comment]
@lbelinsk please let me know if there is more to do in wix-ui related to the `<Box />` component.
for example, where any additional documentation is needed

**Comment by barakplasma** [comment]
talked with @lbelinsk, and I'm going to change the api a bit to allow an end user to change any of these props
- spacing
- vertical alignment
- cross axis alignment

_without_ touching the theme

**Comment by barakplasma** [comment]
https://github.com/wix/wix-ui/pull/126/commits/fc7210bd2f674a1c54effbf1ef1fc68387e77667 is a proposal to include a prop for making the last element in the box take the remaining width.

Notice the difference between the "horizontal" box and the "label..." box.
screen shot 2018-01-02 at 17 52 04

**Comment by barakplasma** [comment]
P.S. https://github.com/wix/wix-ui/pull/126/commits/fc7210bd2f674a1c54effbf1ef1fc68387e77667#diff-7593a963782c9cbb38bf31feee9fbaffL42 shows that we had to change the storybook usage slightly to use the vertical prop properly

File: packages/wix-ui-core/src/components/Box/Box.driver.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+export const boxDriverFactory = ({element, componentInstance}: BoxDriverFactoryInput) => {
+  return {
+    /** check if element exists */
+    exists: () => !!element,
+    /** return box flex direction value */
+    getFlexDirection: () => window.getComputedStyle(element).flexDirection,
+    /** return box item alignment value */
+    getAlignment: () => window.getComputedStyle(element).alignItems,
+    getChildStyle: (child: number): CSSStyleDeclaration => window.getComputedStyle(element.children[child])
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/Box.driver.ts]: missing doc for this line

File: packages/wix-ui-core/src/components/Box/Box.driver.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+export interface BoxDriverFactoryInput {
+  element: HTMLElement;
+  componentInstance: any;
+}
+
+export const boxDriverFactory = ({element, componentInstance}: BoxDriverFactoryInput) => {
+  return {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/Box.driver.ts]: can drop the return in favor of ({ }) syntax

File: packages/wix-ui-core/src/components/Box/styles.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
+import {horizontalCore, verticalCore, BoxTheme} from './theme';
+import * as defaultsDeep from 'lodash/defaultsDeep';
+
+// any <Box /> which isn't vertical will deafult to horizontal - https://github.com/wix/wix-ui/pull/126#discussion_r159144509
+
+export const styles = (theme: BoxTheme) => {
+  const verticalTheme = defaultsDeep({...theme}, verticalCore);
+  const horizontalTheme = defaultsDeep(theme, horizontalCore);
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/styles.ts]: why different syntax? {...theme} or theme ?

File: packages/wix-ui-core/src/components/Box/Box.tsx [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Box
+ */
+class Box extends React.PureComponent<BoxProps> {
+  static propTypes = {
+    /** vertical variant using flex-direction column rather than row */
+    vertical: bool,
+    /** Last Item Takes Remaining Width boolean is useful for setting an input field to full width */
+    lastItemTakesRemainingWidth: bool,
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/Box.tsx]: @shlomitc wdyt about this additional feature?
Should we keep it simple, or is it ok to have it?

File: packages/wix-ui-core/src/components/Box/Box.tsx [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    /** any nodes to be rendered */
+    children: any,
+    /** the class being applied to make it a flex box with flex children */
+    classes: object.isRequired
+  };
+
+  render() {
+    const {children, classes, vertical, lastItemTakesRemainingWidth} = this.props;
+    const classNames = classnames(classes.boxRoot, {[classes.vertical]: vertical}, {[classes.horizontal]: !vertical}, {[classes.lastItemTakesRemainingWidth]: lastItemTakesRemainingWidth});
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/Box.tsx]: long line :)
Please break it

File: packages/wix-ui-core/src/components/Box/styles.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    },
+    horizontal: {
+      alignItems: crossAxisAlignmentMap[horizontalTheme.crossAxisAlignment],
+      flexDirection: 'row',
+      '& >:not(:last-child)': {
+        marginRight: horizontalTheme.spacing
+      }
+    },
+    lastItemTakesRemainingWidth: {
+      '& *:last-child': {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/styles.ts]: why "*" ?

File: packages/wix-ui-core/src/components/Box/styles.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
+import {horizontalCore, verticalCore, BoxTheme} from './theme';
+import * as defaultsDeep from 'lodash/defaultsDeep';
+
+// any <Box /> which isn't vertical will deafult to horizontal - https://github.com/wix/wix-ui/pull/126#discussion_r159144509
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/styles.ts]: Please remove this comment, the implementation is orrect by having horizontal by default

File: packages/wix-ui-core/src/components/Box/Box.tsx (hunk 11 above)
Comment (barakplasma) [reply packages/wix-ui-core/src/components/Box/Box.tsx]: @shlomitc suggested this idea to me:

File: packages/wix-ui-core/src/components/Box/Box.driver.ts [hunk 15]
```diff
@@ -3,14 +3,14 @@ export interface BoxDriverFactoryInput {
   componentInstance: any;
 }
 
-export const boxDriverFactory = ({element, componentInstance}: BoxDriverFactoryInput) => {
-  return {
+export const boxDriverFactory = ({element, componentInstance}: BoxDriverFactoryInput) => ({
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/Box/Box.driver.ts]: componentInstance is not in use

**Pull request merged**